src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let clones = [state];
+
+  actions.forEach(({ type, extraData, keysToRemove }) => {
+    const lastState = clones[clones.length - 1];
+
+    switch (type) {
+      case 'addProperties':
+        clones = [...clones, {
+          ...lastState,
+          ...extraData,
+        }];
+        break;
+
+      case 'removeProperties':
+        const newState = { ...lastState };
+
+        keysToRemove.forEach(key => {
+          delete newState[key];
+        });
+
+        clones = [
+          ...clones,
+          newState,
+        ];
+        break;
+
+      case 'clear':
+        clones = [
+          ...clones,
+          {},
+        ];
+        break;
+
+      default:
+        throw new Error('Wrong action type!');
+    }
+  });
+
+  clones.shift();
```

Comment:
Really, really strange. Why do we need to create an array with one item and then remove it?